### PR TITLE
Add mx.meshgrid

### PR DESCRIPTION
**Pull request by abeleinin**
## Adds mx.meshgrid support

Implements `mx.meshgrid` (#935) support in C++ and Python.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/ops.cpp
Comment (angeloskath) [inline mlx/ops.cpp:703]: A better choice would be `std::string_view indexing`.

File: mlx/ops.cpp
Comment (angeloskath) [inline mlx/ops.cpp:706]: This should be `if (indexing != "xy" && indexing != "ij")`, no need for vector creation etc.

File: mlx/ops.cpp
Comment (angeloskath) [inline mlx/ops.cpp:701]: This doesn't really make sense in MLX since operations are really always returning a new array. For instance you can't have a "view" of one array that is then updated and the view is updated as well (which is kind of the point for this argument).

File: mlx/ops.cpp
Comment (angeloskath) [inline mlx/ops.cpp:728]: The following can be simplified as

```cpp
// The shape vector from above that can be reused throughout.
// You really only need to create this once.
std::vector<int> shape(ndim, 1);

shape[1] = arrays[0].size();
outputs[0] = reshape(arrays[0], shape);
shape[1] = 1;
shape[0] = arrays[1].size();
outputs[1] = reshape(arrays[1], shape);
```

File: python/tests/test_ops.py
Comment (angeloskath) [inline python/tests/test_ops.py:1790]: Thanks!

File: python/src/ops.cpp
Comment (angeloskath) [inline python/src/ops.cpp:2521]: Should be removed, as mentioned above.

File: mlx/ops.cpp
Comment (angeloskath) [inline mlx/ops.cpp:700]: `const std::vector<array>&`

**Review by angeloskath — CHANGES_REQUESTED** [review]
Sorry for the slew of comments. This is a very nice PR, thanks for adding that!

The main thing is the removal of `copy` which should also simplify a lot of things in the op.

**Comment by abeleinin** [comment]
Thanks for the comments. Let me know if anything else needs to be adjusted.

File: mlx/ops.cpp
Comment (awni) [inline mlx/ops.cpp:714]:
```suggestion
    outputs.push_back(reshape(arrays[i], std::move(shape), s));
```

File: mlx/ops.cpp
Comment (awni) [inline mlx/ops.cpp:724]:
```suggestion
    outputs[1] = reshape(arrays[1], std::move(shape));
```

File: mlx/ops.cpp
Comment (awni) [inline mlx/ops.cpp:702]: Let's use a `std::string` here solely for the sake of MLX C (to save binding a new type).

Comment (awni) [reply mlx/ops.cpp:702]: It might work with `string_view` (@andresy )?

File: mlx/ops.cpp
Comment (awni) [inline mlx/ops.cpp:724]: Reshapes need to be done on the correct stream.

File: mlx/ops.cpp
Comment (awni) [reply mlx/ops.cpp:702]: @abeleinin do you mind using `std::string` here?

Comment (awni) [reply mlx/ops.cpp:702]: At @abeleinin we should use a `std::string` here, do you mind updating it?

**Review by awni — APPROVED** [review]
Thanks! LGTM! We can merge once @angeloskath approves and tests clear.

**Comment by abeleinin** [comment]
Sorry for any delays. Glad to get this merged!

**Review by angeloskath — APPROVED** [review]

**Pull request merged**
